core/src/plan/validate.rs

```diff
@@ -41,5 +47,81 @@ pub fn validate(schema_map: &HashMap<String, Schema>, statement: Statement) -> R
         }
     }
 
-    Ok(statement)
+    Ok(())
+}
+
+type SchemaContext<'a> = HashMap<(&'a String, &'a Option<TableAlias>), &'a Schema>;
+type SchemaMap = HashMap<String, Schema>;
+
+pub fn update_schema_context<'a>(
+    schema_map: &'a SchemaMap,
+    statement: &'a Statement,
+) -> SchemaContext<'a> {
+    match statement {
+        Statement::Query(query) => update_schema_context_by_query(schema_map, query),
+        Statement::Insert {
+            table_name, source, ..
+        } => {
+            let table_schema = schema_map
+                .get(table_name)
+                .map(|schema| HashMap::from([((table_name, &None), schema)]))
+                .unwrap_or_else(HashMap::new);
+            let source_schema_list = update_schema_context_by_query(schema_map, source);
+
+            table_schema.into_iter().chain(source_schema_list).collect()
+        }
+        Statement::DropTable { names, .. } => names
+            .iter()
+            .flat_map(|name| {
+                let schema = schema_map.get(name);
+                schema
+                    .map(|schema| HashMap::from([((name, &None), schema)]))
+                    .unwrap_or_else(HashMap::new)
+            })
+            .collect(),
+        _ => HashMap::new(),
+    }
+}
+
+fn update_schema_context_by_query<'a>(
+    schema_map: &'a SchemaMap,
+    query: &'a Query,
+) -> SchemaContext<'a> {
+    let Query { body, .. } = query;
+    match body {
+        SetExpr::Select(select) => {
+            let TableWithJoins { relation, joins } = &select.from;
+
+            let by_table = match relation {
+                TableFactor::Table { name, alias, .. } => {
+                    let schema = schema_map.get(name);
+                    schema
+                        .map(|schema| HashMap::from([((name, alias), schema)]))
+                        .unwrap_or_default()
+                }
+                TableFactor::Derived { subquery, .. } => {
+                    update_schema_context_by_query(schema_map, subquery)
+                }
+                TableFactor::Series { .. } | TableFactor::Dictionary { .. } => HashMap::new(),
+            };
+
+            let by_joins = joins
+                .iter()
+                .flat_map(|Join { relation, .. }| match relation {
+                    TableFactor::Table { name, alias, .. } => {
+                        let schema = schema_map.get(name);
+                        schema
+                            .map(|schema| HashMap::from([((name, alias), schema)]))
+                            .unwrap_or_default()
+                    }
+                    TableFactor::Derived { subquery, .. } => {
+                        update_schema_context_by_query(schema_map, subquery)
+                    }
+                    TableFactor::Series { .. } | TableFactor::Dictionary { .. } => HashMap::new(),
+                });
+
```

Comment:
```suggestion
    let foo = move |table_factor: &'a TableFactor| match table_factor {
        TableFactor::Table { name, alias, .. } => {
            let schema = schema_map.get(name);
            schema
                .map(|schema| HashMap::from([((name, alias), schema)]))
                .unwrap_or_default()
        }
        TableFactor::Derived { subquery, .. } => {
            update_schema_context_by_query(schema_map, subquery)
        }
        TableFactor::Series { .. } | TableFactor::Dictionary { .. } => HashMap::new(),
    };

    match body {
        SetExpr::Select(select) => {
            let TableWithJoins { relation, joins } = &select.from;

            let by_table = foo(relation);

            let by_joins = joins.iter().flat_map(|Join { relation, .. }| foo(relation));

```

Would you consider adding a closure?